providers/ms_entra_id.go

```diff
@@ -0,0 +1,195 @@
+		multiTenantAllowedTenants: opts.MicrosoftEntraIDConfig.AllowedTenants,
+		microsoftGraphURL:         microsoftGraphURL,
+	}
+}
+
+// EnrichSession checks for group overage after calling generic EnrichSession
+func (p *MicrosoftEntraIDProvider) EnrichSession(ctx context.Context, session *sessions.SessionState) error {
+	if err := p.OIDCProvider.EnrichSession(ctx, session); err != nil {
+		return fmt.Errorf("unable to enrich session: %v", err)
+	}
+
+	hasGroupOverage, err := p.checkGroupOverage(session)
+	if err != nil {
+		return fmt.Errorf("unable to check token: %v", err)
+	}
+
+	if hasGroupOverage {
+		logger.Printf("entra overage found, reading groups from Graph API")
+		if err = p.addGraphGroupsToSession(ctx, session); err != nil {
+			return fmt.Errorf("unable to enrich session: %v", err)
+		}
+	} else {
+		logger.Printf("entra group overage not found")
+	}
+
+	return nil
+}
+
+// ValidateSession checks for allowed tenants (e.g. for multi-tenant apps) and passes through to generic ValidateSession
+func (p *MicrosoftEntraIDProvider) ValidateSession(ctx context.Context, session *sessions.SessionState) bool {
+
+	if p.isMultiTenant {
+		tenant, err := p.getTenantFromToken(session)
+		if err != nil {
+			logger.Errorf("unable to retrieve entra tenant from token: %v", err)
+			return false
+		}
+		logger.Errorf("entra multi-tenant: issuer verified successfully, tenant: %s", tenant)
+	}
+
+	if len(p.multiTenantAllowedTenants) > 0 {
+		tenant, err := p.getTenantFromToken(session)
+		if err != nil {
+			logger.Errorf("unable to retrieve entra tenant from token: %v", err)
+			return false
+		}
+
+		tenantAllowed := p.checkTenantMatchesTenantList(tenant, p.multiTenantAllowedTenants)
+		if !tenantAllowed {
+			logger.Printf("entra: tenant %s is not specified in the list of allowed tenants", tenant)
+			return false
+		}
+		logger.Printf("entra: tenant %s is allowed", tenant)
+	}
+
+	return p.OIDCProvider.ValidateSession(ctx, session)
+}
+
+// checkGroupOverage checks ID token's group membership claims for the group overage
+func (p *MicrosoftEntraIDProvider) checkGroupOverage(session *sessions.SessionState) (bool, error) {
+	extractor, err := p.getClaimExtractor(session.IDToken, session.AccessToken)
+	if err != nil {
+		return false, fmt.Errorf("unable to get claim extractor: %v", err)
+	}
+
+	claimNames, _, _ := extractor.GetClaim("_claim_names")
+
+	var claimNamesMap map[string]string
+
+	if claimNames != nil {
+		claimNamesMap = cast.ToStringMapString(claimNames)
+	}
+
+	if _, groupsKeyPresent := claimNamesMap["groups"]; groupsKeyPresent {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (p *MicrosoftEntraIDProvider) addGraphGroupsToSession(ctx context.Context, s *sessions.SessionState) error {
+	groupsHeaders := makeAuthorizationHeader(tokenTypeBearer, s.AccessToken, nil)
+	groupsHeaders.Add("ConsistencyLevel", "eventual")
+
+	groupsURL := fmt.Sprintf("%s/transitiveMemberOf?$select=id&$top=999", p.microsoftGraphURL)
```

Comment:
https://learn.microsoft.com/en-us/graph/api/orgcontact-list-transitivememberof?view=graph-rest-1.0&tabs=http#optional-query-parameters

The default and maximum page sizes are 100 and 999 objects respectively.

Reply:
@jjlakis do we need to introduce pagination here? Because there could be more than 999 groups.

For proper pagination support. We would need to do one call using `"%s/transitiveMemberOf?$count` and then do multiple calls like so:

```go
for skip = 0; skip < count; skip = skip+100 {
  call("%s/transitiveMemberOf?$skip=%d&$top=%d", skip, skip+100)
}
```

Reply:
Found documentation on how paging works. The skip parameter works but the following is recommend:
https://learn.microsoft.com/en-us/graph/paging?view=graph-rest-1.0&tabs=http

Reply:
Already implemented a solution based on the docs. Adding two test right now.

Reply:
Paging is now added including unit testing